In [1]:
import pandas as pd
import numpy as np
from datetime import datetime
import matplotlib.pyplot as plt

In [3]:
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)

## Reading the data

In [9]:
data = pd.read_csv('data/pollution_us_2000_2016.csv')

## Some elementary analysis

In [10]:
data.columns

Index(['Unnamed: 0', 'State Code', 'County Code', 'Site Num', 'Address', 'State', 'County', 'City', 'Date Local', 'NO2 Units', 'NO2 Mean', 'NO2 1st Max Value', 'NO2 1st Max Hour', 'NO2 AQI', 'O3 Units', 'O3 Mean', 'O3 1st Max Value', 'O3 1st Max Hour', 'O3 AQI', 'SO2 Units', 'SO2 Mean', 'SO2 1st Max Value', 'SO2 1st Max Hour', 'SO2 AQI', 'CO Units', 'CO Mean', 'CO 1st Max Value', 'CO 1st Max Hour', 'CO AQI'], dtype='object')

In [11]:
data.head()

Unnamed: 0  State Code  County Code  Site Num                                  Address    State    County     City  Date Local          NO2 Units   NO2 Mean  NO2 1st Max Value  NO2 1st Max Hour  NO2 AQI           O3 Units   O3 Mean  O3 1st Max Value  O3 1st Max Hour  O3 AQI          SO2 Units  SO2 Mean  SO2 1st Max Value  SO2 1st Max Hour  SO2 AQI           CO Units   CO Mean  CO 1st Max Value  CO 1st Max Hour  CO AQI
0           0           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix  2000-01-01  Parts per billion  19.041667               49.0                19       46  Parts per million  0.022500             0.040               10      34  Parts per billion  3.000000                9.0                21     13.0  Parts per million  1.145833               4.2               21     NaN
1           1           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix  2000-01-01  Parts per billion  19.041667               49.0                19       46  Parts per million  0.022500             0.040               10      34  Parts per billion  3.000000                9.0                21     13.0  Parts per million  0.878947               2.2               23    25.0
2           2           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix  2000-01-01  Parts per billion  19.041667               49.0                19       46  Parts per million  0.022500             0.040               10      34  Parts per billion  2.975000                6.6                23      NaN  Parts per million  1.145833               4.2               21     NaN
3           3           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix  2000-01-01  Parts per billion  19.041667               49.0                19       46  Parts per million  0.022500             0.040               10      34  Parts per billion  2.975000                6.6                23      NaN  Parts per million  0.878947               2.2               23    25.0
4           4           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix  2000-01-02  Parts per billion  22.958333               36.0                19       34  Parts per million  0.013375             0.032               10      27  Parts per billion  1.958333                3.0                22      4.0  Parts per million  0.850000               1.6               23     NaN

In [12]:
data.nunique()

Unnamed: 0           134576
State Code               47
County Code              73
Site Num                110
Address                 204
State                    47
County                  133
City                    144
Date Local             5996
NO2 Units                 1
NO2 Mean              31859
NO2 1st Max Value       990
NO2 1st Max Hour         24
NO2 AQI                 129
O3 Units                  1
O3 Mean                8196
O3 1st Max Value        134
O3 1st Max Hour          24
O3 AQI                  125
SO2 Units                 1
SO2 Mean              12736
SO2 1st Max Value       921
SO2 1st Max Hour         24
SO2 AQI                 140
CO Units                  1
CO Mean               34123
CO 1st Max Value       2698
CO 1st Max Hour          24
CO AQI                  107
dtype: int64

#### Observing the stats about data up till here, I found the follwoing:
    - I should drop Unnamed: 0. This is just an index column.
    
    - The NO2 units, SO2 units, O3 units and CO units have no significance.
        - They only have a single unique value which is "Parts per billion"
        - I will drop these too as they would require a lot of memory.

--- billion vs million

In [15]:
data = data.drop(['Unnamed: 0','NO2 Units','O3 Units','SO2 Units','CO Units'],axis=1)

In [16]:
data.head()

State Code  County Code  Site Num                                  Address    State    County     City  Date Local   NO2 Mean  NO2 1st Max Value  NO2 1st Max Hour  NO2 AQI   O3 Mean  O3 1st Max Value  O3 1st Max Hour  O3 AQI  SO2 Mean  SO2 1st Max Value  SO2 1st Max Hour  SO2 AQI   CO Mean  CO 1st Max Value  CO 1st Max Hour  CO AQI
0           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix  2000-01-01  19.041667               49.0                19       46  0.022500             0.040               10      34  3.000000                9.0                21     13.0  1.145833               4.2               21     NaN
1           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix  2000-01-01  19.041667               49.0                19       46  0.022500             0.040               10      34  3.000000                9.0                21     13.0  0.878947               2.2               23    25.0
2           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix  2000-01-01  19.041667               49.0                19       46  0.022500             0.040               10      34  2.975000                6.6                23      NaN  1.145833               4.2               21     NaN
3           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix  2000-01-01  19.041667               49.0                19       46  0.022500             0.040               10      34  2.975000                6.6                23      NaN  0.878947               2.2               23    25.0
4           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix  2000-01-02  22.958333               36.0                19       34  0.013375             0.032               10      27  1.958333                3.0                22      4.0  0.850000               1.6               23     NaN

#### I also feel that I should drop the following columns, but I am keping those for now
    - Country Code, State Code, Site Num
        - These are just identifiers and have no significance.
        - But for now I am keeping these because I think we can use them as id selectors for selections using d3.
        
    - Address
        - This is the address of the recording station.
        - I am keeping this as I think we can pin point these locations on the map and hovering over them would get you an infobox with the details of the selected recording station.
        - But this will be an optional feature!

## Now, let's check NaN values in the dataset

In [17]:
data.isnull().sum()

State Code                0
County Code               0
Site Num                  0
Address                   0
State                     0
County                    0
City                      0
Date Local                0
NO2 Mean                  0
NO2 1st Max Value         0
NO2 1st Max Hour          0
NO2 AQI                   0
O3 Mean                   0
O3 1st Max Value          0
O3 1st Max Hour           0
O3 AQI                    0
SO2 Mean                  0
SO2 1st Max Value         0
SO2 1st Max Hour          0
SO2 AQI              872907
CO Mean                   0
CO 1st Max Value          0
CO 1st Max Hour           0
CO AQI               873323
dtype: int64

* That's strange! Out of the four AQI columns, SO2 AQI and CO AQI have around 872k null values!
* But I am not going to blindly drop these rows!
* Because that would be a loss of so much of data.
* I'll now come up with a strategy to fill up these values.

#### I have noticed another strange thing about this dataset
* There are 4 rows for each day for each City. 
* It seems as if the data recording had some problem.
* One of the rows has no None values.
* But the values for SO2 and CO statistics are different for them!
* Maybe these are updated in steps! Idk. I'll have to research on this.
* Then I can keep the most recently updated value and drop the others.
* For now, I am assuming that this is the correct row and I'll "blindly" drop the other three rows.

In [18]:
data[(data['Date Local'] == '2000-01-01') & (data['City'] == 'Phoenix')]

State Code  County Code  Site Num                                  Address    State    County     City  Date Local   NO2 Mean  NO2 1st Max Value  NO2 1st Max Hour  NO2 AQI  O3 Mean  O3 1st Max Value  O3 1st Max Hour  O3 AQI  SO2 Mean  SO2 1st Max Value  SO2 1st Max Hour  SO2 AQI   CO Mean  CO 1st Max Value  CO 1st Max Hour  CO AQI
0           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix  2000-01-01  19.041667               49.0                19       46   0.0225              0.04               10      34     3.000                9.0                21     13.0  1.145833               4.2               21     NaN
1           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix  2000-01-01  19.041667               49.0                19       46   0.0225              0.04               10      34     3.000                9.0                21     13.0  0.878947               2.2               23    25.0
2           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix  2000-01-01  19.041667               49.0                19       46   0.0225              0.04               10      34     2.975                6.6                23      NaN  1.145833               4.2               21     NaN
3           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix  2000-01-01  19.041667               49.0                19       46   0.0225              0.04               10      34     2.975                6.6                23      NaN  0.878947               2.2               23    25.0

In [19]:
data.shape

(1746661, 24)

The original no of rows are 1746661. By my logic they shoudld be close to 1746661/4 = 436665.25  after dropping Nan values.

In [20]:
data = data.dropna() 

In [21]:
data.shape

(436876, 24)

### Mexico data?

There is data for Mexico too! But we're not interested in that. So, probably I am going to drop the mexico rows.

In [22]:
data = data[data['State'] != 'Country Of Mexico']

In [23]:
data.shape

(434493, 24)

## Rolling Up

In [38]:
(data.shape[0]/30)

14483.1

In [34]:
len(data['County'].value_counts())

132

In [31]:
data[(data['City'] == 'Phoenix')]

State Code  County Code  Site Num                                  Address    State    County     City  Date Local   NO2 Mean  NO2 1st Max Value  NO2 1st Max Hour  NO2 AQI   O3 Mean  O3 1st Max Value  O3 1st Max Hour  O3 AQI  SO2 Mean  SO2 1st Max Value  SO2 1st Max Hour  SO2 AQI   CO Mean  CO 1st Max Value  CO 1st Max Hour  CO AQI
1                 4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix  2000-01-01  19.041667               49.0                19       46  0.022500             0.040               10      34  3.000000                9.0                21     13.0  0.878947               2.2               23    25.0
5                 4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix  2000-01-02  22.958333               36.0                19       34  0.013375             0.032               10      27  1.958333                3.0                22      4.0  1.066667               2.3                0    26.0
9                 4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix  2000-01-03  38.125000               51.0                 8       48  0.007958             0.016                9      14  5.250000               11.0                19     16.0  1.762500               2.5                8    28.0
13                4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix  2000-01-04  40.260870               74.0                 8       72  0.014167             0.033                9      28  7.083333               16.0                 8     23.0  1.829167               3.0               23    34.0
17                4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix  2000-01-05  48.450000               61.0                22       58  0.006667             0.012                9      10  8.708333               15.0                 7     21.0  2.700000               3.7                2    42.0
...             ...          ...       ...                                      ...      ...       ...      ...         ...        ...                ...               ...      ...       ...               ...              ...     ...       ...                ...               ...      ...       ...               ...              ...     ...
1722950           4           13      9997                       4530 N 17TH AVENUE  Arizona  Maricopa  Phoenix  2016-03-22   9.383333               31.4                 7       29  0.037792             0.048               10      44  0.075000                0.8                 7      0.0  0.300000               0.4                7     5.0
1722954           4           13      9997                       4530 N 17TH AVENUE  Arizona  Maricopa  Phoenix  2016-03-23   7.945455               30.2                23       28  0.029042             0.042                8      39  0.004348                0.1                22      0.0  0.250000               0.4               23     5.0
1722958           4           13      9997                       4530 N 17TH AVENUE  Arizona  Maricopa  Phoenix  2016-03-24  20.195833               41.0                21       39  0.022625             0.049               11      45  0.220833                1.1                 9      1.0  0.458333               0.6                5     7.0
1722962           4           13      9997                       4530 N 17TH AVENUE  Arizona  Maricopa  Phoenix  2016-03-25  17.308333               40.2                22       38  0.027125             0.051               11      47  0.445833                1.0                 0      1.0  0.470833               0.8                1     9.0
1722966           4           13      9997                       4530 N 17TH AVENUE  Arizona  Maricopa  Phoenix  2016-03-26  14.579167               35.2                 0       33  0.042053             0.060               11      67  0

In [28]:
data.groupby('Date Local').count()

State Code  County Code  Site Num  Address  State  County  City  NO2 Mean  NO2 1st Max Value  NO2 1st Max Hour  NO2 AQI  O3 Mean  O3 1st Max Value  O3 1st Max Hour  O3 AQI  SO2 Mean  SO2 1st Max Value  SO2 1st Max Hour  SO2 AQI  CO Mean  CO 1st Max Value  CO 1st Max Hour  CO AQI
Date Local                                                                                                                                                                                                                                                                                         
2000-01-01          39           39        39       39     39      39    39        39                 39                39       39       39                39               39      39        39                 39                39       39       39                39               39      39
2000-01-02          38           38        38       38     38      38    38        38                 38                38       38       38                38               38      38        38                 38                38       38       38                38               38      38
2000-01-03          38           38        38       38     38      38    38        38                 38                38       38       38                38               38      38        38                 38                38       38       38                38               38      38
2000-01-04          40           40        40       40     40      40    40        40                 40                40       40       40                40               40      40        40                 40                40       40       40                40               40      40
2000-01-05          40           40        40       40     40      40    40        40                 40                40       40       40                40               40      40        40                 40                40       40       40                40               40      40
...                ...          ...       ...      ...    ...     ...   ...       ...                ...               ...      ...      ...               ...              ...     ...       ...                ...               ...      ...      ...               ...              ...     ...
2016-05-27           7            7         7        7      7       7     7         7                  7                 7        7        7                 7                7       7         7                  7                 7        7        7                 7                7       7
2016-05-28           7            7         7        7      7       7     7         7                  7                 7        7        7                 7                7       7         7                  7                 7        7        7                 7                7       7
2016-05-29           7            7         7        7      7       7     7         7                  7                 7        7        7                 7                7       7         7                  7                 7        7        7                 7                7       7
2016-05-30           7            7         7        7      7       7     7         7                  7                 7        7        7                 7                7       7         7                  7                 7        7        7                 7                7       7
2016-05-31           7            7         7        7      7       7     7         7                  7                 7        7        7                 7                7       7         7                  7                 7        7        7                 7                7       7

[5996 rows x 23 columns]